src/internal/sdata/csv.go

```diff
@@ -174,3 +174,58 @@ func (p *CSVParser) Next(row Tuple) error {
 	}
 	return nil
 }
+
+type HeaderCSVParser struct {
+	CSVParser
+	fields       []string
+	headerRead   bool
+	fieldIndices map[int]int // maps read column to tuple column
+}
+
+func NewHeaderCSVParser(r io.Reader, fields []string) TupleReader {
```

Comment:
I think it's probably better to push this logic into the `CSVParser`, because we have some existing tests that test only `CSVParser` right now.